src/Futhark/Fmt/Monad.hs

```diff
@@ -314,143 +311,140 @@ hardline = do
+hardStdIndent :: Fmt -> Fmt
 hardStdIndent = hardIndent 2
 
 -- | Idents with the standard size of two.
-stdIndent :: (Format a) => a -> FmtM Fmt
+stdIndent :: Fmt -> Fmt
 stdIndent = indent 2
 
 -- | Creates a piece of text, it should not contain any new lines.
-text :: T.Text -> FmtM Fmt
+text :: T.Text -> Fmt
 text t = do
   modify (\s -> s {lastOutput = Just Text})
   pure $ P.pretty t
 
 -- | Adds brackets.
-brackets :: (Format a) => a -> FmtM Fmt
-brackets a = text "[" <:> fmt a <:> text "]"
+brackets :: Fmt -> Fmt
+brackets a = text "[" <:> a <:> text "]"
 
 -- | Adds braces.
-braces :: (Format a) => a -> FmtM Fmt
-braces a = text "{" <:> fmt a <:> text "}"
+braces :: Fmt -> Fmt
+braces a = text "{" <:> a <:> text "}"
 
 -- | Add parenthesis.
-parens :: (Format a) => a -> FmtM Fmt
-parens a = text "(" <:> fmt a <:> text ")"
+parens :: Fmt -> Fmt
+parens a = text "(" <:> a <:> text ")"
 
 -- | Depending on if @b@ is multiline then add a line between them and an indent
 -- to @b@. If singleline then just seperate by a single space.
-(<+/>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(<+/>) :: (Format b, Located b) => Fmt -> b -> Fmt
 (<+/>) a b =
   case lineLayout b of
-    Just MultiLine -> a </> hardStdIndent b
-    Just SingleLine -> a <+> b
+    Just MultiLine -> a </> hardStdIndent (fmt b)
+    Just SingleLine -> a <+> fmt b
     Nothing -> do
       lo <- ask
       case lo of
-        MultiLine -> a </> hardStdIndent b
-        SingleLine -> a <+> b
+        MultiLine -> a </> hardStdIndent (fmt b)
+        SingleLine -> a <+> fmt b
 
 -- | If in a singleline layout then concatenate with 'nil' and in multiline
 -- concatenate by a line.
-(<:/>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(<:/>) :: Fmt -> Fmt -> Fmt
 a <:/> b = a <:> (nil <|> hardline) <:> b
 
 -- | Concatenates @a@ and @b@.
-(<:>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(<:>) :: Fmt -> Fmt -> Fmt
 a <:> b = do
   c <- fmtComments a
-  a' <- fmt a
+  a' <- a
   setTrailingComment a
-  b' <- fmt b
+  b' <- b
   pure $ c <> a' <> b'
 
 -- | Concatenate with a space between.
-(<+>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(<+>) :: Fmt -> Fmt -> Fmt
 a <+> b = a <:> space <:> b
 
 -- | Concatenate with a space if in singleline layout and concatenate by a
 -- line in multiline.
-(</>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(</>) :: Fmt -> Fmt -> Fmt
 a </> b = a <:> line <:> b
 
 -- | If in a singleline layout then choose @a@, if in a multiline layout choose
 -- @b@.
-(<|>) :: (Format a, Format b) => a -> b -> FmtM Fmt
+(<|>) :: Fmt -> Fmt -> Fmt
 a <|> b = do
   lo <- ask
   if lo == SingleLine
-    then fmt a
-    else fmt b
+    then a
+    else b
 
 -- | If in singleline layout seperate by spaces. In a multiline layout seperate
 -- by a single line if two neighbouring elements are singleline. Otherwise
 -- sepereate by two lines.
-sepDecs ::
-  (Format a) =>
-  [a] ->
-  FmtM Fmt
+sepDecs :: [Fmt] -> Fmt
```

Comment:
sepDecs also needs to know the location of each element in order to know if there should be inserted a single line or two lines between declarations.